RTC shell:  shell command group for RTC access

KimBP opened the pull request:
The very first start of a shell command group for RTC access

bjarki-trackunit commented:
Hi, first impressions look very good :)

The CI is failing because you need to specify that the sample you have added should be built for boards which have an RTC, this is achieved by adding these two lines to your samples .yaml
```
tests:
  sample.drivers.rtc:
    tags:
      - RTC
    filter: dt_alias_exists("rtc")
    depends_on: rtc
```

KimBP commented:
> Hi, first impressions look very good :)
> 
> The CI is failing because you need to specify that the sample you have added should be built for boards which have an RTC, this is achieved by adding these two lines to your sample's .yaml
> 
> ```
> tests:
>   sample.drivers.rtc:
>     tags:
>       - RTC
>     filter: dt_alias_exists("rtc")
>     depends_on: rtc
> ```

@bjarki-trackunit : Twister keeps failing due to timeout when running the sample project for native_posix. But a sample project is typically supposed to run forever, so what am I missing in respect to force twister not to timeout - or at least not regard the timeout as an error?

bjarki-trackunit commented:
I see, you need to add a board which is buildable, but not testable, since we don't actually want to run the sample, just verify that it builds. To do this, add
```
    ...
    depends_on: rtc
    integration_platforms:
      - frdm_k64f
```

The frdm_k64f board is not emulatable, and will only be built, contrary to the natix_posix board which is used for unit testing, and automatically included if integration_platforms has not been specified :)

gautierg-st commented on drivers/rtc/rtc_ll_stm32.c:
```diff
@@ -0,0 +1,229 @@
+/* clang-format on */
+
+#if defined(CONFIG_SOC_SERIES_STM32F1X)
+#error RTC not supported for STM32F1X
```
From what I can see the F1 series uses different functions as all other series, but still has an RTC. I can understand that this first version of the driver has no support for F1, but I wonder what is the impact of this. If a user is trying to use the RTC the old way (as a counter) does this error occur? If so, that is a problem.
(Furthermore, there would be plenty of other compilation errors beside this one, since many LL functions used here don't exist on F1.)

KimBP replied:
I do understand your confusion. I see I'm missing a commit, where COUNTER and RTC are made mutual exclusive

gautierg-st commented on drivers/rtc/rtc_ll_stm32.c:
```diff
@@ -0,0 +1,229 @@
+#if defined(CONFIG_SOC_SERIES_STM32H7X) && defined(CONFIG_CPU_CORTEX_M4)
+	LL_C2_EXTI_EnableIT_0_31(RTC_EXTI_LINE);
+	LL_EXTI_EnableRisingTrig_0_31(RTC_EXTI_LINE);
+#elif defined(CONFIG_SOC_SERIES_STM32U5X)
```
This could be `#elif !defined(CONFIG_SOC_SERIES_STM32U5X)`
Or you could do a `#if defined(RTC_EXTI_LINE)` before the H7 part to exclude all series that don't have an exti line (currently only U5, but there could be more in the future).

KimBP replied:
Removed. External interrupt not relevant to driver

gautierg-st commented on drivers/rtc/rtc_ll_stm32.c:
```diff
@@ -0,0 +1,229 @@
+static const struct stm32_pclken rtc_clk[] = {STM32_CLOCK_INFO(0, DT_DRV_INST(0)),
+/* Use Kconfig to configure source clocks fields (Deprecated) */
+/* Fortunately, values are consistent across enabled series */
+#ifdef CONFIG_COUNTER_RTC_STM32_CLOCK_LSE
```
This config was for the RTC used as a counter. It has been kept for compatibility reasons but the new preferred way is to define a source clock in dts.
Since this is a new driver, I think we should just support the new way and display an error if a source clock is not provided.

KimBP replied:
Removed.

gautierg-st reviewed (dismissed):
Reviewed mostly the STM32 driver part. A few points to address but otherwise looks good. Thanks

erwango commented:
@KimBP to ease the review (and hopefully getting this merged faster), it would be better to split the PR in 2:
STM32 RTC
RTC Shell

It will help getting right reviewers on the right parts and avoid shell part blocking the stm32 driver part (and vice versa).

KimBP commented on drivers/rtc/rtc_ll_stm32.c:
```diff
@@ -0,0 +1,229 @@
+	|| defined(CONFIG_SOC_SERIES_STM32H7X) \
+	|| defined(CONFIG_SOC_SERIES_STM32H5X) \
+	|| defined(CONFIG_SOC_SERIES_STM32WLX)
+#define RTC_EXTI_LINE	LL_EXTI_LINE_17
```
Properly correct. Will look into that. It was partly stolen from the counter RTC variant, which is properly a bit old no

KimBP replied:
Interrupts not really used by the driver now, so have removed that part

henrikbrixandersen reviewed (changes requested):
Please split this PR in two: one for adding the STM32 RTC, one for adding the RTC shell.

Furthermore, I'd like to see a real integration test for the added RTC shell module using e.g. a fake RTC. You may be able to find inspiration in how the CAN shell module is tested here: https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/can/shell

KimBP commented:
@erwango It was actually two PR's initially. You suggested to fix my sample issue with e.g. frdm_k64f, but since no new RTC support exists for NXP I thought by joining the two PR's I could make it work for an ST.
Still though I do get a twister error I can't explain. 

I'll split the PR into two again

henrikbrixandersen commented:
> @erwango It was actually two PR's initially. You suggested to fix my sample issue with e.g. frdm_k64f, but since no new RTC support exists for NXP I thought by joining the two PR's I could make it work for an ST. Still though I do get a twister error I can't explain.
> 
> I'll split the PR into two again

You will need to write a hardware-independent test suite for the RTC shell. Please see my previous comment for inspiration.

KimBP commented:
Separate PR #59779 made for driver @gautierg-st, @henrikbrixandersen

Prevas-KimB committed:
drivers: rtc: rtc_shell: rtc shell command suite

For now clock can be read and written.

Clock can be set in full iso8601 time format, like 2023-12-24T12:45:56
or by just providing either data or time

Alarms not accessible

Signed-off-by: Kim Bøndergaard <kim.bondergaard@prevas.dk>

Prevas-KimB committed:
drivers: rtc: rtc_fake: fff rtc driver added

Can be valuable for unit testing modules accessing the RTC

Signed-off-by: Kim Bøndergaard <kim.bondergaard@prevas.dk>

Prevas-KimB committed:
test: drivers: rtc: shell: Unittest of rtc_shell

The test mocks an rtc using the rtc_fake driver

Signed-off-by: Kim Bøndergaard <kim.bondergaard@prevas.dk>

KimBP commented:
driver part removed as requested.
sample project removed and replaced with test project using a rtc_fake driver

henrikbrixandersen commented on drivers/rtc/rtc_shell.c:
```diff
@@ -0,0 +1,234 @@
+
+static int cmd_set(const struct shell *sh, size_t argc, char **argv)
+{
+	const struct device *dev = DEVICE_DT_GET(DT_ALIAS(rtc));
```
The RTC shell commands should be able to work on any RTC device in the system, not jus the one defined via an alias. Check the other shell modules for reference on how to let the user specify the device.

KimBP replied:
Fixed. It becomes a bit more troublesome to call the commands though. I guess it is very uncommon to have more than one RTC device. Would be nice if the alias could be used.

For my stm board it will now be 

`rtc get rtc@44007800
`

I'll have to run 'device list' first to find the right device name

bjarki-trackunit replied:
Given that more than one RTC is quite uncommon, I would be OK with using the alias for now, keeping it simple, until we get a better solution for iterating through RTCs, like the one that exists for sensors, or maybe the new `zephyr,rtc` which may become relevant if this PR #57800 is merged :)

henrikbrixandersen reviewed (changes requested):
Minor nit, otherwise looks good.

bjarki-andreasen reviewed (dismissed):
Very nice work! I played around with it a bit on native_posix and it worked as expected :)

bjarki-trackunit commented:
@erwango @gautierg-st @henrikbrixandersen Would you kindly re-review this PR?

erwango commented on tests/drivers/rtc/shell/src/main.c:
```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright (c) 2022 Vestas Wind Systems A/S
```
Please fix copyrights

bjarki-trackunit commented:
Let me know when the PR is ready for review again :)

bjarki-trackunit commented:
@cfriedt Could you please review this one :)

KimBP commented:
Can I do something to give this PR attention again?
What about the 2 changes requested? I think they are both fixed, but can't see how to get rid of the message

bjarki-trackunit commented:
I will bring it up at tomorrows dev-review to get some attention to it, and ping @erwango @henrikbrixandersen :)

erwango reviewed (approved):
Tested OK on STM32U5
